File: src/web/admin/shared/utils/cambodiaIdentity.ts
/**
 * Cambodia identity formats — keep in sync with backend core/cambodia_identity.py
 *
 * License and plate share one numbered form: 2TE-1507 (NLL-NNNN)
 * Leading digits are the province registration code (1–25).
 */

export const PLATE_FORMAT_LABEL = 'NLL-NNNN';
export const PLATE_FORMAT_EXAMPLE = '2TE-1507';
export const PLATE_FORMAT_REGEX = /^[1-9]\d?[A-Z]{1,2}-\d{4}$/;

/** License follows plate number form across all modules. */
export const LICENSE_FORMAT_LABEL = PLATE_FORMAT_LABEL;
export const LICENSE_FORMAT_EXAMPLE = PLATE_FORMAT_EXAMPLE;
export const LICENSE_FORMAT_REGEX = PLATE_FORMAT_REGEX;

/** Official Cambodia plate province codes (1–25). */
export const CAMBODIA_PROVINCE_CODES = new Set([
  '1', '2', '3', '4', '5', '6', '7', '8', '9', '10',
  '11', '12', '13', '14', '15', '16', '17', '18', '19', '20',
  '21', '22', '23', '24', '25',
]);

export const DEFAULT_VEHICLE_IDENTITY_CONFIG = {
  plate_format: PLATE_FORMAT_LABEL,
  plate_format_example: PLATE_FORMAT_EXAMPLE,
  license_format: LICENSE_FORMAT_LABEL,
  license_format_example: LICENSE_FORMAT_EXAMPLE,
};

/** Normalize to dashed Cambodia private plate / license: 2TE-1507 */
export function formatCambodiaPlate(raw: string): string {
  const cleaned = (raw || '').toUpperCase().replace(/[^A-Z0-9]/g, '');
  if (!cleaned) return '';
  const m = cleaned.match(/^(\d{1,2})([A-Z]{0,2})(\d{0,4})$/);
  if (m) {
    const [, dig, letters, nums] = m;
    if (!letters && !nums) return dig;
    if (!nums) return `${dig}${letters}`;
    return `${dig}${letters}-${nums}`;
  }
  if (cleaned.length > 4) {
    return `${cleaned.slice(0, -4)}-${cleaned.slice(-4)}`;
  }
  return cleaned;
}

export function isValidCambodiaPlate(plate: string): boolean {
  return PLATE_FORMAT_REGEX.test(formatCambodiaPlate(plate));
}

/**
 * Infer province code from plate leading digits.
 * Prefers 2-digit codes (12 = Phnom Penh) over 1-digit when both match.
 * Example: 2TE-1507 → "2" (Battambang); 12AA-1234 → "12" (Phnom Penh).
 */
export function inferProvinceCodeFromPlate(raw: string): string | null {
  const cleaned = (raw || '').toUpperCase().replace(/[^A-Z0-9]/g, '');
  if (!cleaned) return null;
  const m = cleaned.match(/^(\d{1,2})/);
  if (!m) return null;
  const digits = m[1];
  if (digits.length >= 2 && CAMBODIA_PROVINCE_CODES.has(digits.slice(0, 2))) {
    return digits.slice(0, 2);
  }
  if (CAMBODIA_PROVINCE_CODES.has(digits[0])) {
    return digits[0];
  }
  return null;
}

/** License uses the same numbered form as plates. */
export function formatCambodiaLicense(raw: string): string {
  return formatCambodiaPlate(raw);
}

export function isValidCambodiaLicense(license: string): boolean {
  return isValidCambodiaPlate(license);
}

File: src/backend/core/cambodia_identity.py
"""Cambodia identity formats shared by API, seeds, and System Settings.

License and plate use the same numbered form:
  2TE-1507  →  ^[1-9]\d?[A-Z]{1,2}-\d{4}$
"""
from __future__ import annotations

import re

PLATE_FORMAT_LABEL = 'NLL-NNNN'
PLATE_FORMAT_EXAMPLE = '2TE-1507'
PLATE_FORMAT_REGEX = r'^[1-9]\d?[A-Z]{1,2}-\d{4}$'

# License follows the same plate-number form across all modules.
LICENSE_FORMAT_LABEL = PLATE_FORMAT_LABEL
LICENSE_FORMAT_EXAMPLE = PLATE_FORMAT_EXAMPLE
LICENSE_FORMAT_REGEX = PLATE_FORMAT_REGEX

PLATE_RE = re.compile(PLATE_FORMAT_REGEX)
LICENSE_RE = re.compile(LICENSE_FORMAT_REGEX)

DEFAULT_VEHICLE_CONFIG = {
    'plate_format': PLATE_FORMAT_LABEL,
    'plate_format_example': PLATE_FORMAT_EXAMPLE,
    'plate_format_regex': PLATE_FORMAT_REGEX,
    'license_format': LICENSE_FORMAT_LABEL,
    'license_format_example': LICENSE_FORMAT_EXAMPLE,
    'license_format_regex': LICENSE_FORMAT_REGEX,
    'require_owner_link': True,
    'unknown_vehicle_alert': True,
    'retention_days': '365',
}


def normalize_plate(value: str | None) -> str:
    """Normalize to dashed Cambodia private plate: 2TE-1507."""
    raw = (value or '').upper().strip()
    if not raw:
        return ''
    cleaned = ''.join(ch for ch in raw if ch.isalnum())
    if not cleaned:
        return ''
    m = re.fullmatch(r'(\d{1,2})([A-Z]{1,2})(\d{4})', cleaned)
    if m:
        return f'{m.group(1)}{m.group(2)}-{m.group(3)}'
    if len(cleaned) > 4:
        return f'{cleaned[:-4]}-{cleaned[-4:]}'
    return cleaned


def is_valid_plate(value: str | None) -> bool:
    return bool(PLATE_RE.fullmatch(normalize_plate(value)))


def normalize_license(value: str | None) -> str:
    """License uses the same form as plate numbers (e.g. 2TE-1507)."""
    return normalize_plate(value)


def is_valid_license(value: str | None) -> bool:
    return is_valid_plate(value)
